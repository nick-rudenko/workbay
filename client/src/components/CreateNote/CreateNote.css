.create-note {
  width: 500px;
  background: #e3e1e8;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  position: relative;
}

.create-note__input {
  min-height: 40px;
  height: auto;
  border-radius: 8px;
  border: none;
  padding: 8px 16px;
  color: #000;
  background: #e3e1e8;
}

.create-note__input--title {
  font-size: 28px;
  font-weight: bold;
  line-height: 42px;
}

.create-note__input--initial {
  cursor: text;
}

.create-note__input--body {
  font-size: 16px;
  line-height: 24px;
}

.create-note__input:focus {
  outline: none;
}

.create-note__input[contentEditable='true']:empty:before {
  content: attr(data-placeholder);
  color: #8e90a6;
}

.create-note__footer {
  padding: 8px 16px;
  display: flex;
}

.create-note__save {
  height: 32px;
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  padding: 0 16px;
  background: #3366ff;
  border-radius: 4px;
  color: #fff;
  font-size: 14px;
  font-weight: 500;
}
