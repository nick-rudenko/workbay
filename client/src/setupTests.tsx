// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom/extend-expect';
import React, { ComponentType } from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from '@reduxjs/toolkit';
import rootReducer from './store/rootReducer';

const customRender = (ui: React.ReactElement<any>, options?: any) => {
  // Wrap dispatch in a mock so it can be spied on.
  const store = createStore(rootReducer, options?.state);
  store.dispatch = jest.fn(store.dispatch);

  const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
    return <Provider store={store}>{children}</Provider>;
  };

  const returns = render(ui, {
    wrapper: AllTheProviders as ComponentType,
    ...options,
  });

  return { store, ...returns };
};

export * from '@testing-library/react';
export { customRender as render };
