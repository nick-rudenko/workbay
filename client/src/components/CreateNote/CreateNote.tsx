import React, {
  useState,
  SyntheticEvent,
  useRef,
  memo,
  useLayoutEffect,
} from 'react';
import { useDispatch } from 'react-redux';

import { notesActions } from '../../features/notes';
import useOuterClick from '../../hooks/useOuterClick';
import './CreateNote.css';

const initialNoteContent = { title: '', body: '' };

const AddNote = () => {
  const dispatch = useDispatch();
  const [noteContent, setNoteContent] = useState(initialNoteContent);
  const [isExpanded, setIsExpanded] = useState(false);

  const formRef = useOuterClick(saveNote);
  const titleRef = useRef<HTMLDivElement>(null);
  const bodyRef = useRef<HTMLDivElement>(null);

  function expand() {
    setIsExpanded(true);
  }

  useLayoutEffect(() => {
    if (bodyRef.current) bodyRef.current.focus();
  }, [isExpanded]);

  function cleanupAfterSave() {
    if (titleRef?.current?.innerText) titleRef.current.innerText = '';
    if (bodyRef?.current?.innerText) bodyRef.current.innerText = '';
  }

  function saveNote() {
    if (noteContent.title || noteContent.body) {
      sendNote();
      cleanupAfterSave();
    }
    setIsExpanded(false);
  }

  function sendNote() {
    dispatch(notesActions.createNoteRequest(noteContent));
    setNoteContent(initialNoteContent);
  }

  function updateInputValue(e: SyntheticEvent) {
    const targetElement = e.currentTarget as HTMLElement;
    const inputType = targetElement.dataset.type as string;

    setNoteContent({
      ...noteContent,
      [inputType]: targetElement.textContent,
    });
  }

  return (
    <div className="create-note" ref={formRef}>
      {isExpanded ? (
        <>
          <div
            key="active-title"
            data-placeholder="Title"
            data-type="title"
            className="create-note__input create-note__input--title"
            ref={titleRef}
            onInput={updateInputValue}
            contentEditable
            suppressContentEditableWarning
          />

          <div
            key="active-body"
            ref={bodyRef}
            onInput={updateInputValue}
            data-type="body"
            className="create-note__input create-note__input--body"
            data-placeholder="Take a note..."
            contentEditable
            suppressContentEditableWarning
          />

          <div className="create-note__footer">
            <div className="create-note__save" onClick={saveNote}>
              Save
            </div>
          </div>
        </>
      ) : (
        <div
          key="initial"
          onClick={expand}
          onFocus={expand}
          className="create-note__input create-note__input--body create-note__input--initial"
          data-placeholder="✍️ Take a note..."
          contentEditable
          suppressContentEditableWarning
        />
      )}
    </div>
  );
};

export default memo(AddNote);
