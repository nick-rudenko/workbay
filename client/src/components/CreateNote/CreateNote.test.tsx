import React from 'react';
import { act } from 'react-dom/test-utils';
import { fireEvent, render, screen } from '../../setupTests';
import CreateNote from './CreateNote';

describe('CreateNote', () => {
  test('renders correctly', () => {
    const { container } = render(<CreateNote />);
    expect(container).toMatchSnapshot();
  });

  test('click expands form', () => {
    render(<CreateNote />);

    act(() => {
      fireEvent.click(screen.getByRole('textbox'));
    });

    expect(screen.getAllByRole('textbox')).toHaveLength(2);
    expect(screen.getByText('Save')).toBeInTheDocument();
  });

  test('focus expands form', () => {
    render(<CreateNote />);

    act(() => {
      fireEvent.focus(screen.getByRole('textbox'));
    });

    expect(screen.getAllByRole('textbox')).toHaveLength(2);
    expect(screen.getByText('Save')).toBeInTheDocument();
  });

  test('click outside collapses expanded form', () => {
    render(<CreateNote />);

    act(() => {
      fireEvent.focus(screen.getByRole('textbox'));
      fireEvent.click(document);
    });

    expect(screen.getAllByRole('textbox')).toHaveLength(1);
    // use queryBy absence assertions
    expect(screen.queryByText('Save')).not.toBeInTheDocument();
  });

  // testing-library relies on JSDom which doesn't support
  // modifications of contentEditable elements
  test('saves empty note', async () => {
    const { store } = render(<CreateNote />);

    await act(async () => {
      await fireEvent.focus(screen.getByRole('textbox'));
      await fireEvent.click(screen.getByRole('button'));
    });

    expect(store.dispatch).toHaveBeenCalledWith({
      type: 'notes/createNoteRequest',
      payload: {
        title: '',
        body: '',
      },
    });
  });
});
