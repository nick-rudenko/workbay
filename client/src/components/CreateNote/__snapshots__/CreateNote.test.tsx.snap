// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNote renders correctly 1`] = `
<div>
  <div
    class="create-note"
    role="form"
  >
    <div
      class="create-note__input create-note__input--body create-note__input--initial"
      contenteditable="true"
      data-placeholder="✍️ Take a note..."
      role="textbox"
    />
  </div>
</div>
`;
